File: nimapp1/game.nim
# 推箱子核心逻辑（Nim 教学）

import std/[sets, strutils, sequtils]

type
  Hist = object
    px, py: int
    boxFrom, boxTo: string
    isPush: bool
  GameState* = object
    walls, goals, boxes: HashSet[string]
    px*, py*: int
    moves*: int
    won*: bool
    width*, height*: int
    hist: seq[Hist]

proc key(x, y: int): string = $x & "," & $y

proc fromRows*(rows: openArray[string], index = 0): GameState =
  var s: GameState
  var maxX, maxY = 0
  for y, row in rows:
    maxY = y
    for x, ch in row:
      if x > maxX: maxX = x
      let k = key(x, y)
      case ch
      of '#': s.walls.incl k
      of '.': s.goals.incl k
      of '$': s.boxes.incl k
      of '*':
        s.boxes.incl k
        s.goals.incl k
      of '@':
        s.px = x; s.py = y
      of '+':
        s.px = x; s.py = y
        s.goals.incl k
      else: discard
  s.width = maxX + 1
  s.height = maxY + 1
  result = s

proc checkWin(s: var GameState) =
  s.won = true
  for b in s.boxes:
    if b notin s.goals:
      s.won = false
      return

proc tryMove*(s: var GameState, dx, dy: int): bool =
  if s.won: return false
  let nx = s.px + dx
  let ny = s.py + dy
  let nk = key(nx, ny)
  if nk in s.walls: return false
  if nk in s.boxes:
    let bx = nx + dx
    let by = ny + dy
    let bk = key(bx, by)
    if bk in s.walls or bk in s.boxes: return false
    s.hist.add Hist(px: s.px, py: s.py, boxFrom: nk, boxTo: bk, isPush: true)
    s.boxes.excl nk
    s.boxes.incl bk
    s.px = nx; s.py = ny
    inc s.moves
    s.checkWin()
    return true
  s.hist.add Hist(px: s.px, py: s.py, isPush: false)
  s.px = nx; s.py = ny
  true

proc undo*(s: var GameState): bool =
  if s.won or s.hist.len == 0: return false
  while s.hist.len > 0:
    let e = s.hist.pop()
    if e.isPush:
      s.px = e.px; s.py = e.py
      s.boxes.excl e.boxTo
      s.boxes.incl e.boxFrom
      if s.moves > 0: dec s.moves
      s.won = false
      return true
    s.px = e.px; s.py = e.py
  true

proc renderAscii*(s: GameState): string =
  for y in 0 ..< s.height:
    for x in 0 ..< s.width:
      let k = key(x, y)
      if s.px == x and s.py == y:
        result.add(if k in s.goals: '+' else: '@')
      elif k in s.boxes:
        result.add(if k in s.goals: '*' else: '$')
      elif k in s.walls:
        result.add '#'
      elif k in s.goals:
        result.add '.'
      else:
        result.add ' '
    result.add '\n'

File: winformsapp1/Game.cs
// 推箱子核心（WinForms / WPF / Avalonia / TUI 共用教学逻辑可复制）
using System.Collections.Generic;
using System.Text;

namespace SokobanGui;

public sealed class GameState
{
    public HashSet<string> Walls { get; } = new();
    public HashSet<string> Goals { get; } = new();
    public HashSet<string> Boxes { get; } = new();
    public (int X, int Y) Player { get; private set; }
    public int Moves { get; private set; }
    public bool Won { get; private set; }
    public int Width { get; private set; }
    public int Height { get; private set; }
    private readonly List<Hist> _hist = new();
    private readonly record struct Hist((int X, int Y) Player, string? BoxFrom, string? BoxTo);

    public static string Key(int x, int y) => $"{x},{y}";

    public static GameState FromRows(string[] rows)
    {
        var s = new GameState();
        int maxX = 0, maxY = 0;
        for (int y = 0; y < rows.Length; y++)
        {
            maxY = y;
            var row = rows[y];
            for (int x = 0; x < row.Length; x++)
            {
                if (x > maxX) maxX = x;
                var k = Key(x, y);
                switch (row[x])
                {
                    case '#': s.Walls.Add(k); break;
                    case '.': s.Goals.Add(k); break;
                    case '$': s.Boxes.Add(k); break;
                    case '*': s.Boxes.Add(k); s.Goals.Add(k); break;
                    case '@': s.Player = (x, y); break;
                    case '+': s.Player = (x, y); s.Goals.Add(k); break;
                }
            }
        }
        s.Width = maxX + 1;
        s.Height = maxY + 1;
        return s;
    }

    public bool TryMove(int dx, int dy)
    {
        if (Won) return false;
        var (px, py) = Player;
        int nx = px + dx, ny = py + dy;
        var nk = Key(nx, ny);
        if (Walls.Contains(nk)) return false;
        if (Boxes.Contains(nk))
        {
            int bx = nx + dx, by = ny + dy;
            var bk = Key(bx, by);
            if (Walls.Contains(bk) || Boxes.Contains(bk)) return false;
            _hist.Add(new Hist(Player, nk, bk));
            Boxes.Remove(nk);
            Boxes.Add(bk);
            Player = (nx, ny);
            Moves++;
            Won = true;
            foreach (var b in Boxes)
                if (!Goals.Contains(b)) { Won = false; break; }
            return true;
        }
        _hist.Add(new Hist(Player, null, null));
        Player = (nx, ny);
        return true;
    }

    public bool Undo()
    {
        if (Won || _hist.Count == 0) return false;
        Hist? e = null;
        while (_hist.Count > 0)
        {
            e = _hist[_hist.Count - 1];
            _hist.RemoveAt(_hist.Count - 1);
            if (e.Value.BoxFrom != null) break;
            Player = e.Value.Player;
        }
        if (e is null || e.Value.BoxFrom is null) return true;
        Player = e.Value.Player;
        Boxes.Remove(e.Value.BoxTo!);
        Boxes.Add(e.Value.BoxFrom);
        if (Moves > 0) Moves--;
        Won = false;
        return true;
    }

    public string RenderAscii()
    {
        var sb = new StringBuilder();
        for (int y = 0; y < Height; y++)
        {
            for (int x = 0; x < Width; x++)
            {
                var k = Key(x, y);
                if (Player == (x, y)) sb.Append(Goals.Contains(k) ? '+' : '@');
                else if (Boxes.Contains(k)) sb.Append(Goals.Contains(k) ? '*' : '$');
                else if (Walls.Contains(k)) sb.Append('#');
                else if (Goals.Contains(k)) sb.Append('.');
                else sb.Append(' ');
            }
            sb.AppendLine();
        }
        return sb.ToString();
    }
}

public static class Levels
{
    public static readonly string[] Mini =
    {
        "#######",
        "#. . .#",
        "# $$$ #",
        "#.$@$.#",
        "# $$$ #",
        "#. . .#",
        "#######",
    };
}
